src/run.rs

```diff
@@ -266,22 +267,50 @@ pub fn clean_up(
     tmp_dir: TempDir,
     worktree_branch: Option<&str>,
     subprocesses: &mut [Child],
+    prompt: bool,
 ) -> Result<()> {
-    if !delete_file.exists() {
+    let mut delete = delete_file.exists();
+
+    if prompt && delete {
+        println!("Are you sure you want to delete this project? (Y/n)");
+        loop {
+            let mut input = String::new();
+
+            match stdin().read_line(&mut input) {
+                Ok(_n) => match input.trim() {
+                    "" | "Yes" | "yes" | "Y" | "y" => break,
```

Comment:
Probably best to still assign delete to true even if it's will always be true. Again if someone change the logic and the thing is false for some reason, it will feel like a trap

Reply:
Actually I think we got this all wrong. Rust allows us to do better

Reply:
Let me introduce you something cool: **compile check of variable initialization.**

```
let delete;
```

At first it looks like a simple variable but it's actually more than that. Let's see what happens if you compile:

```rust
fn main() {
    let delete;
    proceed(delete);
}

fn proceed(value: bool) {
    println!("proceed with: {}", value);
}
```

It will fail to compile and you will get this error:

```
error[E0381]: use of possibly-uninitialized variable: `delete`
 --> src/main.rs:3:13
  |
3 |     proceed(delete);
  |             ^^^^^^ use of possibly-uninitialized `delete`

For more information about this error, try `rustc --explain E0381`.
```

Creating uninitialized variable is actually useful to ensure that a variable is assigned at some point but also that it is assigned only once.

For example, this will fail to compile too:

```rust
fn main() {
    let delete;
    delete = true;
    delete = false;
    proceed(delete);
}

fn proceed(value: bool) {
    println!("proceed with: {}", value);
}
```

```
error[E0384]: cannot assign twice to immutable variable `delete`
 --> src/main.rs:4:5
  |
2 |     let delete;
  |         ------ help: consider making this binding mutable: `mut delete`
3 |     delete = true;
  |     ------------- first assignment to `delete`
4 |     delete = false;
  |     ^^^^^^^^^^^^^^ cannot assign twice to immutable variable

For more information about this error, try `rustc --explain E0384`.
```

The only way for this code to compile is to assign the variable once no matter which path you take. For example, this code will compile:

```rust
fn main() {
    let delete;

    if true {
        delete = true;
    } else {
        delete = false;
    }

    proceed(delete);
}

fn proceed(value: bool) {
    println!("proceed with: {}", value);
}
```

But this code won't:

```rust
fn main() {
    let delete;

    if true {
        delete = true;
    }

    proceed(delete);
}

fn proceed(value: bool) {
    println!("proceed with: {}", value);
}
```

```
error[E0381]: use of possibly-uninitialized variable: `delete`
 --> src/main.rs:8:13
  |
8 |     proceed(delete);
  |             ^^^^^^ use of possibly-uninitialized `delete`
```

This is because the compiler sees that in one branch of the if, the variable won't be assigned.

Reply:
Now the thing with our `delete` variable is that we want one answer at the end: true for deletion, false for preservation, and we don't want to allow any logic to not assign this variable or assign it twice.

```rust
let delete; // not having `mut` here is important because we are not mutating,
            // we are delaying the initialization

if !delete_file.exists() {
    delete = false;
} else if prompt {
    delete = loop {
        // ...
        match input {
            "Yes" => break true,
            "No" => break false,
        }
        // ...
    };
} else {
    delete = true;
}
```

And you can see now that no error is possible.

(I think I'm also introducing you to "breaking with value". Loops in Rust can be written as expression like if-expressions. In which case, you can pass a value after the break keyword and it will yield that value out of the loop)

In our particular case you can also not use delayed initialization at all if you want and use if-expression:

```rust
let delete = if !delete_file.exists() {
    false
} else if prompt {
    loop {
        // ...
        match input {
            "Yes" => break true,
            "No" => break false,
        }
        // ...
    }
} else {
    true
};
```

Whichever you find more readable really... both are better than what we currently have because it enforces the check at compilation, ensuring the logic is correct. If someone wants to add a special case, they will have to make the compiler happy thus making the logic plausible.

Reply:
Hmmm... :thinking: maybe that's what you wanted to tell me there? https://github.com/yozhgoor/cargo-temp/pull/68#discussion_r964745895

Reply:
Damn, what a blogpost :+1: 

> Hmmm... thinking maybe that's what you wanted to tell me there?
Not at all, It was more about how to revert the prompt if we want to delete it later.

Reply:
What about this?

```
let delete = if !delete_file.exists() {
        false
    } else if prompt {
        println!("Are you sure you want to delete this project? (Y/n)");
        loop {
            let mut input = String::new();

            match stdin().read_line(&mut input) {
                Ok(_n) => match input.trim() {
                    "" | "Yes" | "yes" | "Y" | "y" => {
                        break true;
                    }
                    "No" | "no" | "N" | "n" => {
                        break false;
                    }
                    _ => println!("hmm, `{}` doesn't look like `yes` or `no`", input.trim()),
                },
                Err(err) => {
                    log::error!("failed to read input: {}", err);
                }
            }
        }
    } else {
        true
    };
```

I tried with the `let delete;` but my problem was that I can use the value of `delete_file.exists()` instead.
This solution seems the more readable for me and we keep the checks at compile time and the `break with value` parts.

Reply:
> What about this?

That works, that's the second solution I proposed right?

> we keep the checks at compile time

Both solutions does equally.

> I tried with the `let delete;` but my problem was that I can use the value of `delete_file.exists()` instead.

Not sure what you mean sorry... but I think both solutions are equally good. First solution is more flexible in some cases so I hope you learned how it works for the future.